File: code/algo/getoffer/c++/3/check_str_onereplace.cpp
#include <string>
#include <vector>

class Trie {
public:
  Trie *children[26] = {nullptr};
  bool isEnd = false;
};

class MagicDictionary {
public:
  MagicDictionary() { root = new Trie(); }

  void BuildDict(const std::vector<std::string> &dictionary) {
    for (const std::string &word : dictionary) {
      Trie *node = root;
      for (char ch : word) {
        int idx = ch - 'a';
        if (node->children[idx] == nullptr) {
          node->children[idx] = new Trie();
        }
        node = node->children[idx];
      }
      node->isEnd = true;
    }
  }

  bool Search(const std::string &searchWord) {
    return dfs(root, searchWord, 0, false);
  }

private:
  Trie *root;

  bool dfs(Trie *node, const std::string &word, int index, bool modified) {
    if (index == word.size()) {
      return modified && node->isEnd;
    }
    int c = word[index] - 'a';
    if (node->children[c] != nullptr &&
        dfs(node->children[c], word, index + 1, modified)) {
      return true;
    }
    if (!modified) {
      for (int i = 0; i < 26; ++i) {
        if (i != c && node->children[i] != nullptr &&
            dfs(node->children[i], word, index + 1, true)) {
          return true;
        }
      }
    }
    return false;
  }
};

File: code/algo/getoffer/go/3/check_str_onereplace.go
package main

type MagicDictionary struct {
	trie *trie
}

type trie struct {
	children [26]*trie
	isEnd    bool
}

/** Initialize your data structure here. */
func NewMagicDictionary() MagicDictionary {
	return MagicDictionary{trie: &trie{}}
}

func (this *MagicDictionary) BuildDict(dictionary []string) {
	for _, word := range dictionary {
		cur := this.trie
		for _, ch := range word {
			idx := ch - 'a'
			if cur.children[idx] == nil {
				cur.children[idx] = &trie{}
			}
			cur = cur.children[idx]
		}
		cur.isEnd = true
	}
}

func (this *MagicDictionary) Search(searchWord string) bool {
	var dfs func(node *trie, searchWord string, modified bool) bool
	dfs = func(node *trie, searchWord string, modified bool) bool {
		if searchWord == "" {
			return modified && node.isEnd
		}
		c := searchWord[0] - 'a'
		if node.children[c] != nil && dfs(node.children[c], searchWord[1:], modified) {
			return true
		}
		if !modified {
			for i, child := range node.children {
				if i != int(c) && child != nil && dfs(child, searchWord[1:], true) {
					return true
				}
			}
		}
		return false
	}
	return dfs(this.trie, searchWord, false)
}
